### PR TITLE
Add dn resolution function

### PR DESCRIPTION
**Pull request by coffee809721232**
## Summary
Add lib_dn.c to libs/libc/netdb
## Impact
dhcp6c depends on lib_dn.c
## Testing
dhcp6c test passed

### CODE REVIEW

File: libs/libc/netdb/lib_dn.c [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+        {
+          if ((*s & 0xc0) != 0xc0)
+            {
+              return 0;
+            }
+
+          s = base + ((s[0] & 0x3f) << 8 | s[1]);
+        }
+
+      if (!*s)
```
Comment (pkarashchenko) [inline libs/libc/netdb/lib_dn.c:53]:
```suggestion
      if (*s == 0)
```

File: libs/libc/netdb/lib_dn.c [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+static int getlens(FAR unsigned char *lens, FAR const char *s, int l)
+{
+  int i = 0;
+  int j = 0;
+  int k = 0;
+
+  for (; ; )
+    {
+      for (; j < l && s[j] != '.'; j++);
+      if (j - k - 1u > 62)
```
Comment (pkarashchenko) [inline libs/libc/netdb/lib_dn.c:79]: why do we need `1u` and not `1`?

File: libs/libc/netdb/lib_dn.c [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+                 FAR const unsigned char *dn, FAR const char *end,
+                 FAR const unsigned char *lens, int nlen)
+{
+  int l;
+  int o;
+  int m = 0;
+  short offs[128];
+  int noff = getoffs(offs, base, dn);
+
+  if (!noff)
```
Comment (pkarashchenko) [inline libs/libc/netdb/lib_dn.c:106]:
```suggestion
  if (noff == 0)
```

File: libs/libc/netdb/lib_dn.c [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+    {
+      return 0;
+    }
+
+  for (; ; )
+    {
+      l = lens[--nlen];
+      o = offs[--noff];
+      end -= l;
+      if (l != base[o] || memcmp(base + o + 1, end, l))
```
Comment (pkarashchenko) [inline libs/libc/netdb/lib_dn.c:116]:
```suggestion
      if (l != base[o] || (memcmp(base + o + 1, end, l) != 0))
```

File: libs/libc/netdb/lib_dn.c [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+        {
+          return m;
+        }
+
+      *offset = o;
+      m += l;
+      if (nlen)
```
Comment (pkarashchenko) [inline libs/libc/netdb/lib_dn.c:123]:
```suggestion
      if (nlen > 0)
```

File: libs/libc/netdb/lib_dn.c [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+        {
+          m++;
+        }
+
+      if (!nlen || !noff)
```
Comment (pkarashchenko) [inline libs/libc/netdb/lib_dn.c:128]:
```suggestion
      if (nlen == 0 || noff == 0)
```

File: libs/libc/netdb/lib_dn.c [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+  int m = 0;
+  int offset = 0;
+  int bestlen = 0;
+  int bestoff = 0;
+  unsigned char lens[127];
+  FAR unsigned char **p;
+  FAR const char *end;
+  size_t l = strnlen(src, 255);
+
+  if (l && src[l - 1] == '.')
```
Comment (pkarashchenko) [inline libs/libc/netdb/lib_dn.c:156]:
```suggestion
  if (l > 0 && src[l - 1] == '.')
```

File: libs/libc/netdb/lib_dn.c [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+    {
+      l--;
+    }
+
+  if (l > 253 || space <= 0)
+    {
+      return -1;
+    }
+
+  if (!l)
```
Comment (pkarashchenko) [inline libs/libc/netdb/lib_dn.c:166]:
```suggestion
  if (l == 0)
```

File: libs/libc/netdb/lib_dn.c [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+    {
+      *dst = 0;
+      return 1;
+    }
+
+  end = src + l;
+  n = getlens(lens, src, l);
+  if (!n)
```
Comment (pkarashchenko) [inline libs/libc/netdb/lib_dn.c:174]:
```suggestion
  if (n == 0)
```

File: libs/libc/netdb/lib_dn.c [hunk 10]
```diff
@@ -0,0 +1,315 @@
...
+    {
+      return -1;
+    }
+
+  p = dnptrs;
+  if (p && *p)
```
Comment (pkarashchenko) [inline libs/libc/netdb/lib_dn.c:180]:
```suggestion
  if (p != NULL && *p != NULL)
```

File: libs/libc/netdb/lib_dn.c [hunk 11]
```diff
@@ -0,0 +1,315 @@
...
+    {
+      for (p++; *p; p++)
```
Comment (pkarashchenko) [inline libs/libc/netdb/lib_dn.c:182]: Can we rewrite this and above to
```
if (dnptrs != NULL && *dnptrs != NULL)
  {
    for (p = dnptrs + 1; *p != NULL; p++)
  }
```
?

File: libs/libc/netdb/lib_dn.c [hunk 12]
```diff
@@ -0,0 +1,315 @@
...
+
+  memcpy(dst + 1, src, l - bestlen);
+  for (i = j = 0; i < l - bestlen; i += lens[j++] + 1)
+    {
+      dst[i] = lens[j];
+    }
+
+  /* add tail */
+
+  if (bestlen)
```
Comment (pkarashchenko) [inline libs/libc/netdb/lib_dn.c:212]:
```suggestion
  if (bestlen > 0)
```

File: libs/libc/netdb/lib_dn.c [hunk 13]
```diff
@@ -0,0 +1,315 @@
...
+  else
+    {
+      dst[i++] = 0;
+    }
+
+  /* save dst pointer */
+
+  if (i > 2 && lastdnptr && dnptrs && *dnptrs)
+    {
+      while (*p)
```
Comment (pkarashchenko) [inline libs/libc/netdb/lib_dn.c:226]:
```suggestion
      while (*p != NULL)
```

File: libs/libc/netdb/lib_dn.c [hunk 14]
```diff
@@ -0,0 +1,315 @@
...
+      dst[i++] = bestoff;
+    }
+  else
+    {
+      dst[i++] = 0;
+    }
+
+  /* save dst pointer */
+
+  if (i > 2 && lastdnptr && dnptrs && *dnptrs)
```
Comment (pkarashchenko) [inline libs/libc/netdb/lib_dn.c:224]:
```suggestion
  if (i > 2 && lastdnptr != NULL && dnptrs != NULL && *dnptrs != NULL)
```

File: libs/libc/netdb/lib_dn.c [hunk 15]
```diff
@@ -0,0 +1,315 @@
...
+
+  dend = dest + (space > 254 ? 254 : space);
+
+  /* detect reference loop using an iteration counter */
+
+  for (i = 0; i < end - base; i += 2)
+    {
+      /* loop invariants: p<end, dest<dend */
+
+      if (*p & 0xc0)
```
Comment (pkarashchenko) [inline libs/libc/netdb/lib_dn.c:264]:
```suggestion
      if ((*p & 0xc0) != 0)
```

File: libs/libc/netdb/lib_dn.c [hunk 16]
```diff
@@ -0,0 +1,315 @@
...
+/* label start offsets of a compressed domain name s */
+
+static int getoffs(FAR short *offs, FAR const unsigned char *base,
+                   FAR const unsigned char *s)
+{
+  int i = 0;
+
+  for (; ; )
+    {
+      while (*s & 0xc0)
```
[the remark this answers is not in the document]
Comment (coffee809721232) [reply libs/libc/netdb/lib_dn.c]: > 

done

File: libs/libc/netdb/lib_dn.c (hunk 1 above)
Comment (coffee809721232) [reply libs/libc/netdb/lib_dn.c]: done

File: libs/libc/netdb/lib_dn.c (hunk 2 above)
Comment (coffee809721232) [reply libs/libc/netdb/lib_dn.c]: done

File: libs/libc/netdb/lib_dn.c (hunk 3 above)
Comment (coffee809721232) [reply libs/libc/netdb/lib_dn.c]: done

File: libs/libc/netdb/lib_dn.c (hunk 4 above)
Comment (coffee809721232) [reply libs/libc/netdb/lib_dn.c]: done

File: libs/libc/netdb/lib_dn.c (hunk 11 above)
Comment (coffee809721232) [reply libs/libc/netdb/lib_dn.c]: done

File: libs/libc/netdb/lib_dn.c (hunk 12 above)
Comment (coffee809721232) [reply libs/libc/netdb/lib_dn.c]: done

File: libs/libc/netdb/lib_dn.c (hunk 13 above)
Comment (coffee809721232) [reply libs/libc/netdb/lib_dn.c]: done

File: libs/libc/netdb/lib_dn.c (hunk 15 above)
Comment (coffee809721232) [reply libs/libc/netdb/lib_dn.c]: done

File: libs/libc/netdb/lib_dn.c (hunk 14 above)
Comment (coffee809721232) [reply libs/libc/netdb/lib_dn.c]: done

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by pkarashchenko — APPROVED** [review]

File: libs/libc/netdb/lib_dn.c (hunk 11 above)
Comment (pkarashchenko) [reply libs/libc/netdb/lib_dn.c]: sorry, but it is not.

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by pkarashchenko — APPROVED** [review]

**Pull request merged**